File: frontend-app/lib/features/product/data/product_api.dart
import 'package:dio/dio.dart';
import 'package:file_picker/file_picker.dart';
import 'package:shopping_app/core/network/api_client.dart';
import 'package:shopping_app/core/network/api_exception.dart';
import 'package:shopping_app/core/network/api_response.dart';
import 'package:shopping_app/features/product/domain/product_models.dart';

class ProductApi {
  ProductApi(this._client);

  final ApiClient _client;

  Future<List<CategoryNode>> categoryTree() async {
    final response = await _client.get('/api/categories/tree');
    return unwrapApiResponse(
      response.data,
      (value) => readObjectList(value, CategoryNode.fromJson, message: '类目响应格式错误'),
    );
  }

  Future<PageResult<SearchItem>> search({
    String? keyword,
    String? categoryId,
    String sort = 'DEFAULT',
    int page = 1,
    int pageSize = 12,
  }) async {
    final response = await _client.get(
      '/api/search',
      queryParameters: {
        if (keyword != null && keyword.isNotEmpty) 'keyword': keyword,
        if (categoryId != null && categoryId.isNotEmpty) 'categoryId': categoryId,
        'sort': sort,
        'page': page,
        'pageSize': pageSize,
      },
    );
    return unwrapApiResponse(
      response.data,
      (value) => PageResult.fromJson(value, SearchItem.fromJson),
    );
  }

  Future<List<HotWord>> hotWords({int limit = 8}) async {
    final response = await _client.get(
      '/api/search/hot-words',
      queryParameters: {'limit': limit},
    );
    return unwrapApiResponse(response.data, (value) {
      final json = requireJsonMap(value);
      return readObjectList(json['words'], HotWord.fromJson);
    });
  }

  Future<SpuDetail> spuDetail(String spuId) async {
    final response = await _client.get('/api/spu/$spuId');
    return unwrapApiResponse(response.data, SpuDetail.fromJson);
  }

  Future<PageResult<Review>> reviews(String spuId, {int page = 1, int pageSize = 20}) async {
    final response = await _client.get(
      '/api/review/spu/$spuId',
      queryParameters: {'page': page, 'pageSize': pageSize},
    );
    return unwrapApiResponse(
      response.data,
      (value) => PageResult.fromJson(value, Review.fromJson),
    );
  }

  Future<ReviewStats> reviewStats(String spuId) async {
    final response = await _client.get('/api/review/spu/$spuId/stats');
    return unwrapApiResponse(response.data, ReviewStats.fromJson);
  }

  Future<PageResult<SpuItem>> merchantPage({
    String? keyword,
    String? shelf,
    int page = 1,
    int pageSize = 20,
  }) async {
    final response = await _client.get(
      '/api/merchant/spu/page',
      queryParameters: {
        if (keyword != null && keyword.isNotEmpty) 'keyword': keyword,
        if (shelf != null && shelf.isNotEmpty) 'shelf': shelf,
        'page': page,
        'pageSize': pageSize,
      },
    );
    return unwrapApiResponse(
      response.data,
      (value) => PageResult.fromJson(value, SpuItem.fromJson),
    );
  }

  Future<SpuDetail> merchantDetail(String spuId) async {
    final response = await _client.get('/api/merchant/spu/$spuId');
    return unwrapApiResponse(response.data, SpuDetail.fromJson);
  }

  Future<SpuDetail> createSpu({
    required String categoryId,
    required String name,
    String? brand,
    String? mainImage,
    required List<Map<String, Object?>> skus,
  }) async {
    final response = await _client.post(
      '/api/merchant/spu',
      data: {
        'categoryId': categoryId,
        'name': name,
        if (brand != null && brand.isNotEmpty) 'brand': brand,
        if (mainImage != null && mainImage.isNotEmpty) 'mainImage': mainImage,
        'skus': skus,
      },
    );
    return unwrapApiResponse(response.data, SpuDetail.fromJson);
  }

  Future<void> changeStatus(String spuId, String action, {String? remark}) async {
    final response = await _client.put(
      '/api/merchant/spu/$spuId/status',
      data: {'action': action, if (remark != null) 'remark': remark},
    );
    unwrapApiResponse<void>(response.data, (_) {});
  }

  Future<void> adjustStock(String skuId, int change, {String? remark}) async {
    final response = await _client.put(
      '/api/merchant/sku/$skuId/stock',
      data: {'change': change, if (remark != null) 'remark': remark},
    );
    unwrapApiResponse<void>(response.data, (_) {});
  }

  Future<ProductMedia> uploadMedia(PlatformFile file) async {
    final path = file.path;
    if (path == null || path.isEmpty) {
      throw const ApiException(message: '无法读取所选图片');
    }
    final form = FormData.fromMap({
      'file': await MultipartFile.fromFile(path, filename: file.name),
    });
    final response = await _client.postMultipart('/api/merchant/product-media', form);
    return unwrapApiResponse(response.data, ProductMedia.fromJson);
  }
}

File: frontend-miniapp/miniprogram/features/product/data/product-api.ts
import { apiBaseUrl } from '../../../config/environment'
import { apiRequest } from '../../../core/http/api-client'
import { ApiError, type ApiEnvelope } from '../../../core/models/api'
import { readSession } from '../../../core/storage/session-storage'
import type {
  CategoryNode,
  Page,
  Review,
  ReviewStats,
  SearchItem,
  Sku,
  SpuCreateRequest,
  SpuDetail,
  SpuItem,
} from '../domain/product-models'

function withQuery(path: string, params: Record<string, string | number | undefined>): string {
  const parts: string[] = []
  for (const [key, value] of Object.entries(params)) {
    if (value === undefined || value === '') continue
    parts.push(`${encodeURIComponent(key)}=${encodeURIComponent(String(value))}`)
  }
  return parts.length ? `${path}?${parts.join('&')}` : path
}

export const categoryApi = {
  tree(): Promise<CategoryNode[]> {
    return apiRequest({ path: '/api/categories/tree', authenticated: false })
  },
}

export const searchApi = {
  search(params: {
    keyword?: string
    categoryId?: string
    brand?: string
    priceMin?: number
    priceMax?: number
    sort?: string
    page: number
    pageSize: number
  }): Promise<Page<SearchItem>> {
    return apiRequest({
      path: withQuery('/api/search', params),
      authenticated: false,
    })
  },

  hotWords(limit = 10): Promise<{ words: { keyword: string; count: string }[] }> {
    return apiRequest({
      path: withQuery('/api/search/hot-words', { limit }),
      authenticated: false,
    })
  },
}

export const spuApi = {
  detail(spuId: string): Promise<SpuDetail> {
    return apiRequest({ path: `/api/spu/${spuId}`, authenticated: false })
  },

  merchantPage(params: {
    categoryId?: string
    keyword?: string
    status?: string
    shelf?: 'LISTED' | 'UNLISTED'
    page: number
    pageSize: number
  }): Promise<Page<SpuItem>> {
    return apiRequest({ path: withQuery('/api/merchant/spu/page', params) })
  },

  merchantDetail(spuId: string): Promise<SpuDetail> {
    return apiRequest({ path: `/api/merchant/spu/${spuId}` })
  },

  create(payload: SpuCreateRequest): Promise<SpuDetail> {
    return apiRequest({ path: '/api/merchant/spu', method: 'POST', data: payload })
  },

  changeStatus(spuId: string, action: 'SUBMIT' | 'PUBLISH' | 'OFF_SHELF', remark?: string): Promise<SpuItem> {
    return apiRequest({
      path: `/api/merchant/spu/${spuId}/status`,
      method: 'PUT',
      data: { action, remark },
    })
  },

  adjustStock(skuId: string, change: number, remark?: string): Promise<Sku> {
    return apiRequest({
      path: `/api/merchant/sku/${skuId}/stock`,
      method: 'PUT',
      data: { change, remark },
    })
  },
}

export const reviewApi = {
  listBySpu(spuId: string, page = 1, pageSize = 20): Promise<Page<Review>> {
    return apiRequest({
      path: withQuery(`/api/review/spu/${spuId}`, { page, pageSize }),
      authenticated: false,
    })
  },

  stats(spuId: string): Promise<ReviewStats> {
    return apiRequest({ path: `/api/review/spu/${spuId}/stats`, authenticated: false })
  },
}

export const productMediaApi = {
  upload(filePath: string): Promise<{ id: string; url: string; contentType: string }> {
    const accessToken = readSession()?.accessToken
    return new Promise((resolve, reject) => {
      wx.uploadFile({
        url: `${apiBaseUrl()}/api/merchant/product-media`,
        filePath,
        name: 'file',
        header: accessToken ? { Authorization: `Bearer ${accessToken}` } : {},
        success: (response) => {
          try {
            const envelope = JSON.parse(response.data) as Partial<ApiEnvelope<{ id: string; url: string; contentType: string }>>
            if (response.statusCode >= 200 && response.statusCode < 300 && envelope.code === 0 && envelope.data) {
              resolve(envelope.data)
              return
            }
            reject(new ApiError(envelope.message || '上传失败', response.statusCode, envelope.code))
          } catch {
            reject(new ApiError('上传失败', response.statusCode))
          }
        },
        fail: () => reject(new ApiError('无法连接服务器，请检查网络和 API 地址', 0)),
      })
    })
  },
}
